Remora.Rest/Json/DataObjectConverter.cs

```diff
@@ -490,207 +489,92 @@ JsonConverterFactory converterFactory
-            {
-                propertyValue = JsonSerializer.Deserialize(ref reader, propertyType, options);
-            }
-            else
-            {
-                // This converter should only be in effect for the duration of this property; we'll need to clone
-                // the options.
-                var clonedOptions = new JsonSerializerOptions(options);
-                clonedOptions.Converters.Insert(0, converter);
-
-                propertyValue = JsonSerializer.Deserialize(ref reader, propertyType, clonedOptions);
-            }
-
-            // Verify nullability
-            if (!dtoProperty.AllowsNull() && propertyValue is null)
-            {
-                throw new JsonException();
-            }
-
-            if (!readProperties.ContainsKey(dtoProperty))
-            {
-                readProperties.Add(dtoProperty, propertyValue);
-            }
-            else if (isPrimaryChoice)
-            {
-                readProperties[dtoProperty] = propertyValue;
-            }
-
-            if (!reader.Read())
-            {
-                throw new JsonException();
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
             }
         }
 
-        // Reorder and polyfill the read properties
-        var writablePropertyCount = _dtoProperties.Count(p => p.CanWrite);
-        var constructorArguments = new object?[writablePropertyCount];
-        for (var i = 0; i < writablePropertyCount; i++)
-        {
-            var dtoProperty = _dtoProperties[i];
-            if (!readProperties.TryGetValue(dtoProperty, out var propertyValue))
-            {
-                if (dtoProperty.PropertyType.IsOptional())
-                {
-                    propertyValue = _dtoEmptyOptionals[dtoProperty.PropertyType];
-                }
-                else
-                {
-                    throw new JsonException
-                    (
-                        $"The data property \"{dtoProperty.Name}\" did not have a corresponding value in the JSON."
-                    );
-                }
-            }
-
-            constructorArguments[i] = propertyValue;
-        }
+        var converterOptions = new BoundDataObjectConverterOptions<TInterface>
+        (
+            _dtoFactory,
+            _allowExtraProperties,
+            writeProperties.ToArray(),
+            readProperties.ToArray()
+        );
 
-        return _dtoFactory(constructorArguments);
+        return new BoundDataObjectConverter<TInterface, TImplementation>(converterOptions);
     }
 
-    /// <inheritdoc />
-    public override void Write
-    (
-        Utf8JsonWriter writer,
-        TInterface value,
-        JsonSerializerOptions options
-    )
+    private static JsonSerializerOptions CreatePropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
```

Comment:
Is there a reason the converter is nullable here?

Reply:
I could move the null check here:
https://github.com/Remora/Remora.Rest/blob/250f4493b99d857f864dd315e7e7daa5a8d7042b/Remora.Rest/Json/DataObjectConverter.cs#L497-L500
But I felt the code looked cleaner this way.

Reply:
I'd prefer to have the check there, since it's the only place the method is used.

Reply:
Moved the null-check:
https://github.com/Remora/Remora.Rest/blob/916ee44822b0206e371039ab37534decdacdb646/Remora.Rest/Json/DataObjectConverter.cs#L497-L500
https://github.com/Remora/Remora.Rest/blob/916ee44822b0206e371039ab37534decdacdb646/Remora.Rest/Json/DataObjectConverter.cs#L545